bindings/py/cpp_src/bindings/engine/py_Engine.cpp

```diff
@@ -82,6 +83,89 @@ namespace htm_ext
+				/* Request a buffer descriptor from Python */
+				py::buffer_info info = b.request();
+				if (info.ndim != 1)
+					throw std::runtime_error("Expected a one dimensional array!");
+			
+				size_t size = static_cast<size_t>(info.shape[0]);
+				NTA_BasicType type;
+				if     (info.format == py::format_descriptor<Int32>::format()) type = NTA_BasicType_Int32;
+				else if (info.format == py::format_descriptor<UInt32>::format()) type = NTA_BasicType_UInt32;
+				else if (info.format == py::format_descriptor<Int64>::format()) type = NTA_BasicType_Int64;
+				else if (info.format == py::format_descriptor<UInt64>::format()) type = NTA_BasicType_UInt64;
+				else if (info.format == py::format_descriptor<Real32>::format()) type = NTA_BasicType_Real32;
+				else if (info.format == py::format_descriptor<Real64>::format()) type = NTA_BasicType_Real64;
+				else if (info.format == py::format_descriptor<bool>::format()) type = NTA_BasicType_Bool;
+				else if (info.format == py::format_descriptor<Byte>::format()) type = NTA_BasicType_Byte;
+				else throw std::runtime_error("Unexpected data type in the array!");
+				if (copy) {
+					new(&m) Array(type);
+					m.allocateBuffer(size);
+					std::memcpy(m.getBuffer(), info.ptr, size*info.itemsize);
+				}
+				new (&m) Array(type, info.ptr, info.shape[0]);
+			})
+			.def(py::init<>(),           "Create an empty Array object.")
+		    .def("zeroBuffer", &Array::zeroBuffer, "Fills array with zeros")
+			.def("getCount", &Array::getCount, "Returns the number of elements.")
+	
+			// boolean functions to determine the type of value in the Array
+			.def("getType", &Array::getType, "Returns an enum representing the data type.")
+			.def("isInt32",  [](const Array &self) { return self.getType() == NTA_BasicType_Int32; })
+			.def("isUInt32", [](const Array &self) { return self.getType() == NTA_BasicType_UInt32; })
+			.def("isInt64",  [](const Array &self) { return self.getType() == NTA_BasicType_Int64; })
+			.def("isUInt64", [](const Array &self) { return self.getType() == NTA_BasicType_UInt64; })
+			.def("isReal32", [](const Array &self) { return self.getType() == NTA_BasicType_Real32; })
+			.def("isReal64", [](const Array &self) { return self.getType() == NTA_BasicType_Real64; })
+			.def("isBool",   [](const Array &self) { return self.getType() == NTA_BasicType_Bool; })
+			.def("isSDR",    [](const Array &self) { return self.getType() == NTA_BasicType_SDR; })
+			.def(py::pickle(
+				[](const Array& self) {
+					std::stringstream ss;
+					self.save(ss);
+					return py::bytes(ss.str());
+			},
+				[](const py::bytes& s) {
+					std::istringstream ss(s);
+					Array self;
+					self.load(ss);
+					return self;
+			}));
+
+			
+//              py_Array..def(py::init<const SDR &sdr>(), "Create an Array object from an SDR object.",             py::arg("values"));
+//              py_Array.def("getSDR", &Array::getSDR, "Returns an SDR object if the Array contains type SDR.");
```

Comment:
so what is this "Array" object used for? Can't it just be a SDR which has existing bindings?

Reply:
The short answer is no.

In NetworkAPI, the Array class this IS the SDR,    It is a wrapper around an array of any type and contains the type and size of the data.  It can also contain the new SDR class.

What an Array class can do is
-  pass data that is not boolean in nature in addition to the normal SDR boolean data.
- It manages the allocation and deallocation of buffers containing the data.  
- It can also contain external buffers that it does not allocate to avoid copying.